improver/utilities/spatial.py

```diff
@@ -761,3 +761,55 @@ def add_vicinity_coordinate(
         point, units=units, long_name="radius_of_vicinity", attributes=attributes
     )
     cube.add_aux_coord(coord)
+
+
+def maximum_in_height(
+    cube: Cube, lower_height_bound: float = None, upper_height_bound: float = None
+) -> Cube:
+    """Calculate the maximum value over the height coordinate. If bounds are specified
+    then the maximum value between the lower_height_bound and upper_height_bound is calculated.
+
+    If either the upper or lower bound is None then no bound is applied. For example if no
+    lower bound is provided but an upper bound of 300m is provided then the maximum is
+    calculated for all height levels less than 300m.
+
+    Args:
+        cube:
+            A cube with a height coordinate.
+        lower_height_bound:
+            The lower bound for the height coordinate. This is either a float of None if no
+            lower bound is desired. Any specified bounds should have the same units as the
+            height coordinate of cube.
+        upper_height_bound:
+            The upper bound for the height coordinate. This is either a float of None if no
+            upper bound is desired. Any specified bounds should have the same units as the
+            height coordinate of cube.
+    Returns:
+        A cube of the maximum value over the height coordinate or maximum value between the desired
+        height values.
+    """
+    cube_name = cube.name()
+    height_levels = cube.coord("height").points
+
+    # replace None in bounds with a numerical value either below or above the range of height
+    # levels in the cube so it can be used as a constraint.
+    print(lower_height_bound)
+    if lower_height_bound is None:
+        lower_height_bound = min(height_levels)
+    if upper_height_bound is None:
+        upper_height_bound = max(height_levels)
+
+    height_constraint = iris.Constraint(
+        height=lambda height: lower_height_bound <= height <= upper_height_bound
+    )
+    cube_subsetted = cube.extract(height_constraint)
+
+    if len(cube_subsetted.coord("height").points) > 1:
+        max_cube = cube_subsetted.collapsed("height", iris.analysis.MAX)
+    else:
+        max_cube = cube_subsetted
+
+    max_cube.rename(
+        f"maximum_{cube_name}_between_{lower_height_bound}m_and_{upper_height_bound}m"
```

Comment:
The resulting cube has all this in the metadata
- cube.name contains "maximum" and lower and upper height bounds and units.
- cube.coords contains scalar "height" coord with point set to the mid-point and bounds set to the lower and upper height bounds. Coord defines units.
- cube.cell_methods contains info saying this is a maximum over height.

Are we sure we need to change the name too?

Reply:
I'm not sure. I added the bounds into the name to make sure the name was as accurate as possible to what the diagnostic is. It could just be, for example, "Maximum relative humidity" if we're happy that won't be misleading by implying that it is the maximum over all heights when it's only over a small range of heights. As you mention all the information about the range is included in the other metadata.

It also occurs to me that the lower and upper bounds aren't required to be in metres so I'll need to do something to the name whichever way we choose to go.

Reply:
I've asked Bruce for advice here. As for the units of the bounds, they should always be in metres if we are obeying the IMPROVER style guide of always storing data in SI units. I suspect we should therefore enforce SI units here too by stating that the bounds must be in metres and adding a `.convert_units("m")` to the incoming height coord.